SpeechService/Translations.cs

```diff
@@ -748,109 +757,159 @@ public static string Humanize(decimal? value)
             int nextDigit;
             string order;
-            int digits = (int)Math.Log10((double)value);
-            if (digits < 3)
+            int digits = (int)Math.Truncate(Math.Log10((double)value));
+            if (digits < 2)
             {
                 // Units
                 number = (int)value;
                 order = "";
                 nextDigit = (int)((value - number) * 10);
             }
+            else if (digits < 3)
+            {
+                // Hundreds
+                (number, nextDigit) = Normalize(value, 100);
+                order = Properties.Phrases.hundred;
+            }
             else if (digits < 6)
             {
                 // Thousands
-                number = (int)(value / 1E3M);
-                order = " " + Properties.Phrases.thousand;
-                nextDigit = (int)((value - (number * 1E3M)) / 1E2M);
+                (number, nextDigit) = Normalize(value, 1E3M);
+                order = Properties.Phrases.thousand;
             }
             else if (digits < 9)
             {
                 // Millions
-                number = (int)(value / 1E6M);
-                order = " " + Properties.Phrases.million;
-                nextDigit = (int)((value - (number * 1E6M)) / 1E5M);
+                (number, nextDigit) = Normalize(value, 1E6M);
+                order = Properties.Phrases.million;
             }
             else if (digits < 12)
             {
                 // Billions
-                number = (int)(value / 1E9M);
-                order = " " + Properties.Phrases.billion;
-                nextDigit = (int)((value - (number * 1E9M)) / 1E8M);
+                (number, nextDigit) = Normalize(value, 1E9M);
+                order = Properties.Phrases.billion;
             }
             else if (digits < 15)
             {
                 // Trillions
-                number = (int)(value / 1E12M);
-                order = " " + Properties.Phrases.trillion;
-                nextDigit = (int)((value - (number * 1E12M)) / 1E11M);
+                (number, nextDigit) = Normalize(value, 1E12M);
+                order = Properties.Phrases.trillion;
             }
             else
             {
                 // Quadrillions
-                number = (int)(value / 1E15M);
-                order = " " + Properties.Phrases.quadrillion;
-                nextDigit = (int)((value - (number * 1E15M)) / 1E14M);
+                (number, nextDigit) = Normalize(value, 1E15M);
+                order = Properties.Phrases.quadrillion;
             }
 
-            // See if we have an exact match
-            if (((long)(((decimal)value) / (decimal)Math.Pow(10, digits - 1))) * (decimal)(Math.Pow(10, digits - 1)) == value)
+            // emit exact value
+            if (order == "")
             {
-                return minus + number + order;
+                // we recurse only if we know that there _is_ a nested order (i. e. stuff like "51 hundred thousand")
```

Comment:
I'm not sure what you mean by `51 hundred thousand`?

Reply:
See above. If we receive a number in range between `100000` and `1000000`, when we first enter `Humanize()`, we strip the `10^3` and generate a `thousands` suffix, then we recurse to do the same once again to strip the remaining `10^2`.

Same would work with `xxx hundred millions` etc.